### ISSUE TITLE
Clarify whether CacheApi.getOrElseUpdate support null as value or not

### ISSUE THREAD

**Issue by zhengcan** [playframework/playframework#9476]
### Play Version (2.5.x / etc)

Play 2.7.2

### API (Scala / Java / Neither / Both)

Both

### Operating System (Ubuntu 15.10 / MacOS 10.10 / Windows 10)

MacOS Mojave 10.14.5

### JDK (Oracle 1.8.0_72, OpenJDK 1.8.x, Azul Zing)

openjdk version "1.8.0_212"

### Library Dependencies

n/a

### Expected Behavior

I use SyncCacheApi to cache entity. The code looks like:
```java
User findUserById(Long userId) {
  return this.cacheApi.getOrElseUpdate(
    "User_" + userId,
    () -> User.find.query().where().eq("userId", userId").findOne(),
    60
  );
)
```

I expected the function return nullable value as .findOne() work.

### Actual Behavior

Because caffeine will throw NPE to reject null as value, the above code won't work as expected. 

### Suggestion

Please consider to make .getOrElseUpdate support null as value, or optimize document to warn developers.

**Comment by limpid-kzonix** [comment]
It depends on the implementation of cache API.
 Have a look at the internal implementation of cache what you want to use. 
```java
@Nullable
      V put(K key, V value, final long duration, final TimeUnit unit, boolean onlyIfAbsent) {
        Objects.requireNonNull(unit);
        Objects.requireNonNull(value);
        Caffeine.requireArgument(duration >= 0L);
        Expiry<K, V> expiry = new Expiry<K, V>() {
          public long expireAfterCreate(K key, V value, long currentTime) {
            return unit.toNanos(duration);
          }

          public long expireAfterUpdate(K key, V value, long currentTime, long currentDuration) {
            return unit.toNanos(duration);
          }

          public long expireAfterRead(K key, V value, long currentTime, long currentDuration) {
            return currentDuration;
          }
        };
        if (BoundedPolicy.this.cache.isAsync) {
          Expiry<K, V> asyncExpiry = new AsyncExpiry((Expiry)expiry);
          expiry = asyncExpiry;
          V asyncValue = CompletableFuture.completedFuture(value);
          value = asyncValue;
        }

        return BoundedPolicy.this.cache.put(key, value, (Expiry)expiry, true, onlyIfAbsent);
      }
```
e.g with EhCache impl it works "fine" (returns "null").

**Comment by zhengcan** [comment]
I think the adapter should provide the same behavior for different implementations. 
For example, in play.api.cache.caffiene.SyncCaffeineCacheApi, it can check the value and return null instead of Exception when orElse returns null.

**Comment by limpid-kzonix** [comment]
I would propose: 
-  to change API ```scala
def getOrElseUpdate[A: ClassTag](key: String, expiration: Duration = Duration.Inf)(orElse: => Future[A]): Future[A]``` to ```scala
def getOrElseUpdate[A: ClassTag](key: String, expiration: Duration = Duration.Inf)(orElse: => Future[A]): Future[Option[A]]```
since orElse can return a null value.
(and for sync cache api accordingly)
- and to add additional checking of 'orElse' returning value.
 + if 'orElse' returns null -> Option[Nothing]

**Comment by limpid-kzonix** [comment]
Issue is fixed since Play 2.8

**Comment by mkurz** [comment]
> FYI: Issue is fixed since Play 2.8

Closing.
